activate_state.go

```diff
@@ -35,22 +41,78 @@ const (
 // on a successfully unlocked storage container. This is intended to
 // be used like an enum, with the user of the [ActivateContext] API
 // defining the values.
-//
-// XXX: This will eventually be part of the serialized [ActivateState].
 type DeactivationReason string
 
+// KeyslotErrorType describes the reason a keyslot failed when it was attempted.
+type KeyslotErrorType string
+
+const (
+	KeyslotErrorNone                   KeyslotErrorType = ""
+	KeyslotErrorIncompatibleRoleParams KeyslotErrorType = "incompatible-role-params" // The role parameters for the keyslot are not compatible with the current boot configuration.
+	KeyslotErrorInvalidKeyData         KeyslotErrorType = "invalid-key-data"         // The keyslot metadata is invalid.
+	KeyslotErrorInvalidPrimaryKey      KeyslotErrorType = "invalid-primary-key"      // The keyslot's primary key failed the primary key crosscheck.
+	KeyslotErrorIncorrectUserAuth      KeyslotErrorType = "incorrect-user-auth"      // An incorrect user authorization was provided.
+	KeyslotErrorPlatformFailure        KeyslotErrorType = "platform-failure"         // There was an error with the platform device.
+	KeyslotErrorUnknown                KeyslotErrorType = "unknown"
+)
+
+func errorToKeyslotError(err error) KeyslotErrorType {
+	if err == nil {
+		return KeyslotErrorNone
+	}
+
+	if errors.Is(err, errInvalidPrimaryKey) {
+		return KeyslotErrorInvalidPrimaryKey
+	}
+
+	var ikdErr *InvalidKeyDataError
+	if errors.As(err, &ikdErr) {
+		return KeyslotErrorInvalidKeyData
+	}
+
+	var ikdrErr *IncompatibleKeyDataRoleParamsError
+	if errors.As(err, &ikdrErr) {
+		return KeyslotErrorIncompatibleRoleParams
+	}
+
+	// XXX: Add the incorrect PIN error here as well.
+	if errors.Is(err, ErrInvalidPassphrase) || errors.Is(err, errInvalidRecoveryKey) {
+		return KeyslotErrorIncorrectUserAuth
+	}
+
+	var (
+		puErr  *PlatformUninitializedError
+		pduErr *PlatformDeviceUnavailableError
+	)
+	if errors.As(err, &puErr) || errors.As(err, &pduErr) {
+		return KeyslotErrorPlatformFailure
+	}
+
+	return KeyslotErrorUnknown
+}
+
 // ContainerActivateState contains the activation state for a single
 // [StorageContainer].
 type ContainerActivateState struct {
-	Status ActivationStatus `json:"status"`
+	Status           ActivationStatus            `json:"status"`                      // The overall activation status for this storage container.
+	Keyslot          string                      `json:"keyslot,omitempty"`           // If the container was activated, the name of the keyslot used.
+	DeactivateReason DeactivationReason          `json:"deactivate-reason,omitempty"` // An argument supplied to ActivateContext.DeactivateContainer.
+	KeyslotErrors    map[string]KeyslotErrorType `json:"keyslot-errors"`              // A map of errors for tried keyslots, keyed by name.
```

Comment:
Note sure, but maybe that would be interesting to have it ordered. Because if we ever have to report details to users for example, in some advanced reporting mode, then it might be confusing if we cannot show the errors in order of them happening.

Reply:
maybe we should record a list of order we tried the keyslots then?

Reply:
I've added a list of keyslot names ordered by when they failed.